presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -803,6 +804,11 @@ private void transitionToFinished()
     }
 
     public boolean transitionToFailed(Throwable throwable)
+    {
+        return transitionToFailed(throwable, currentState -> currentState != FINISHING && !currentState.isDone());
```

Comment:
Also add comment here why it won't fail the query when query state is `FINISHING`